Library in C++ & Python

RobertLucian opened the pull request:
Here's the old PR for the C++ base library: #331

RobertLucian commented:
1. the issue for the base library + examples in C++: RobertLucian/GrovePi#1 

1. the issue for the library for the sensors/modules written in Python & C++: RobertLucian/GrovePi#5

Shoban94 commented:
Hi Robert, the libraries look great. I have tested the light,color,sound, buzzer, button, rotary angle, leds, dht and ultrasonic sensors. These example programs look great. I think you will want to have the grovepi.h and grovepi.cpp installed system wide or have a copy of them inside folders like grove_dht, so that it is possible for users to compile. Otherwise things look great.

RobertLucian commented on Software/Python/grove_mini_motor_driver/grove_mini_motor_driver.py:
```diff
@@ -0,0 +1,246 @@
+        self.__writeMotor(self.right_motor, self.right_motor.STANDBY, "right motor warning")
+
+"""
+Will be implemented on Monday
```
There will be new functionalities the next week. This will make the life simpler for the guy who wants to use this library.

karan259 commented on Script/install.sh:
```diff
@@ -109,7 +109,7 @@ install_dependencies() {
 	feedback "Installing Dependencies"
 	echo "======================="
-	sudo apt-get install python-pip git libi2c-dev python-serial i2c-tools python-smbus python3-smbus arduino minicom python-dev -y
+	sudo apt-get install python-pip git libi2c-dev python-serial i2c-tools python-smbus python3-smbus python3-scipy arduino minicom python-dev -y
```
why do we need python3-scipy package for

RobertLucian commented:
As @karan259  suggested, I did the following:

1. added `shebangs` for the python sources

1. added licenses for python sources

1. removed `interp1d1` & `scipy` dependency

1. solved unicode syntax error on the `high_temperature_example.py`

1. removed source variables and placed them inside the library (for the `driver_example.py`)

1. used `.out` instead of `.exe` on C++ sources

1. made a `dpkg-deb` package and added installation instructions in the `/Cpp/README.md` - but there's a problem. Please take a look at the contents of the package and tell me if I'm missing anything. The thing is when compiling I still need to add `/usr/local/lib` path in order to work. 

Also, as I have promised one week ago, I added the `README.md` for the C++ library. I also added a README for the RF link library.

Shoban94 commented on Software/Cpp/README.md:
```diff
@@ -32,9 +32,9 @@ See more about the GrovePi here:  https://www.nuget.org/packages/GrovePi/
-$ cd grovepicpplib_0.1
-$ sudo dpkg --install ./grovepicpp_0.1.deb
+$ tar -xzvf grovepicpp_0.1.1.tar.gz
+$ cd grovepicpplib_0.1.1
```
A small typo here `cd grovepicpplib_0.1.1`, it should be `cd grovepicpp_0.1.1`.

RobertLucian replied:
Thanks. It's done now.

Shoban94 commented on Software/Cpp/grove_dht_pro/grove_dht_example.cpp:
```diff
@@ -0,0 +1,89 @@
+using GrovePi::I2CError;
+
+//g++ -Wall grovepi.cpp grove_dht_pro.cpp grove_dht_example.cpp -o grove_dht_example.out -> without package installed
+//g++ -Wall -lgrovepicpp grove_dht_pro.cpp grove_dht_example.cpp -o grove_dht_example.out -> with package installed
```
Will be better to mention the name of the package without version number, so that it will help users understand, which package we are talking about.

Also add sudo to g++ commands as they will throw a permission denied error and many will try to copy paste it directly

RobertLucian replied:
Got it. It's done.

karan259 commented on Software/Python/grove_mini_motor_driver/driver_example.py:
```diff
@@ -0,0 +1,122 @@
+        sys.exit(0)
+
+    # in case there's an IO error aka I2C
+except IOError:l
```
Looks like a syntax error

RobertLucian replied:
Yep, it's out now.

karan259 commented on Software/Python/grove_mini_motor_driver/grove_mini_motor_driver.py:
```diff
@@ -0,0 +1,249 @@
+from scipy.interpolate import interp1d
```
@RobertLucian : Do we need scipy for interpolate.  Also, I am facing a problem when installing scipy on R4R with pip.

No scipy
```
 Traceback (most recent call last):
   File "driver_example.py", line 34, in <module>
     from grove_mini_motor_driver import MiniMotorDriver
   File "/home/pi/Desktop/test/GrovePi/Software/Python/grove_mini_motor_driver/grove_mini_motor_driver.py", line 1, in <module>
     from scipy.interpolate import interp1d
ImportError: No module named scipy.interpolate
```

Problem when installing scipy
```
Command /usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip-build-woKenB/scipy/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-nQjRmB-record/install-record.txt --single-version-externally-managed --compile failed with error code 1 in /tmp/pip-build-woKenB/scipy
Storing debug log for failure in /root/.pip/pip.log

```

RobertLucian replied:
Weird. We've already done that. I've sent a new commit.
It's removed and in its place I've added a custom function which does the same thing.

Cheers!

karan259 commented:
@RobertLucian : I was testing the transmitter example for the RFlink without the module just to see if the program works properly. I am getting this error:
```
pi@dex:~/Desktop/test/GrovePi/Software/Python/grove_rflink433mhz_oneway_kit $ python grove_transmitter_example.py
Traceback (most recent call last):
  File "grove_transmitter_example.py", line 67, in <module>
    Main()
  File "grove_transmitter_example.py", line 57, in Main
    transmitter.writeMessage(message_to_broadcast)
  File "/home/pi/Desktop/test/GrovePi/Software/Python/grove_rflink433mhz_oneway_kit/grove_rflink433mhz.py", line 111, in writeMessage
    self.__writeFragment(message, chr(count), chr(count))
  File "/home/pi/Desktop/test/GrovePi/Software/Python/grove_rflink433mhz_oneway_kit/grove_rflink433mhz.py", line 69, in __writeFragment
    crc_to_str += chr(each_byte + self.crc_offset)
TypeError: cannot concatenate 'str' and 'int' objects
```

karan259 commented:
can you create another section in the readme in the cpp folder and write the steps there to create the tar package and the deb package

karan259 commented on Software/Python/grove_mini_motor_driver/driver_example.py:
```diff
@@ -0,0 +1,122 @@
+        sys.exit(0)
+
+    # in case there's an IO error aka I2C
+except IOError:
```
Add indentation

RobertLucian replied:
It's done!

RobertLucian commented:
Also, I've added `#!/usr/bin/env python3` shebangs so that the user knows we are using `python3` with the 3 libraries I've developed.

DexterInd commented:
Hey @RobertLucian , I checked and I couldn't find any comment codes on the fact that it runs on Python3, maybe we should include a note on that and an example command so folks looking at the example know how to run the example?

RobertLucian commented:
I was thinking that maybe `shebangs` were enough. I'll add some notes in each of the example I've developed.

Apropos, only the libraries I've written shall go with Python 3, because otherwise all the libraries should be tested.
